src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -225,66 +235,227 @@ static void
+	{
+		CacheLocalPlanForTask(task, originalDistributedPlan);
+	}
+	else
+	{
+		EState *executorState = planState->state;
+		ResetExecutionParameters(executorState);
+	}
+}
+
+
+/*
+ * CopyDistributedPlanWithoutCache is a helper function which copies the
+ * distributedPlan into the current memory context.
+ *
+ * We must not change the distributed plan since it may be reused across multiple
+ * executions of a prepared statement. Instead we create a deep copy that we only
+ * use for the current execution.
+ *
+ * We also exclude localPlannedStatements from the copyObject call for performance
+ * reasons, as they are immutable, so no need to have a deep copy.
+ */
+static DistributedPlan *
+CopyDistributedPlanWithoutCache(CitusScanState *scanState)
+{
+	DistributedPlan *originalDistributedPlan = scanState->distributedPlan;
+	List *localPlannedStatements =
+		originalDistributedPlan->workerJob->localPlannedStatements;
+	originalDistributedPlan->workerJob->localPlannedStatements = NIL;
+
+	DistributedPlan *distributedPlan = copyObject(originalDistributedPlan);
+	scanState->distributedPlan = distributedPlan;
+
+	/* set back the immutable field */
+	originalDistributedPlan->workerJob->localPlannedStatements = localPlannedStatements;
+	distributedPlan->workerJob->localPlannedStatements = localPlannedStatements;
+
+	return distributedPlan;
+}
+
+
+/*
+ * ResetExecutionParameters set the parameter list to NULL. See the function
+ * for details.
+ */
+static void
+ResetExecutionParameters(EState *executorState)
+{
+	/*
+	 * We've processed parameters in ExecuteMasterEvaluableFunctions and
+	 * don't need to send their values to workers, since they will be
+	 * represented as constants in the deparsed query. To avoid sending
+	 * parameter values, we set the parameter list to NULL.
+	 */
+	executorState->es_param_list_info = NULL;
+}
+
+
+/*
+ * CacheLocalPlanForTask caches a plan that is local to this node in the
+ * originalDistributedPlan.
+ *
+ * The basic idea is to be able to skip planning on the shards when possible.
+ */
+static void
+CacheLocalPlanForTask(Task *task, DistributedPlan *originalDistributedPlan)
+{
+	PlannedStmt *localPlan = GetCachedLocalPlan(task, originalDistributedPlan);
+	if (localPlan != NULL)
+	{
+		/* we already have a local plan */
+		return;
+	}
+
+	if (list_length(task->relationShardList) == 0)
+	{
+		/* zero shard plan, no need to cache */
+		return;
+	}
+
+	/*
+	 * All memory allocations should happen in the plan's context
+	 * since we'll cache the local plan there.
+	 */
+	MemoryContext oldContext =
+		MemoryContextSwitchTo(GetMemoryChunkContext(originalDistributedPlan));
+
+	/*
+	 * We prefer to use jobQuery (over task->query) because we don't want any
+	 * functions/params have been evaluated in the cached plan.
+	 */
+	Query *shardQuery = copyObject(originalDistributedPlan->workerJob->jobQuery);
+
+	UpdateRelationsToLocalShardTables((Node *) shardQuery, task->relationShardList);
```

Comment:
I kind of wonder whether we should look at whether the shard is local in the existing function, though I'm fine with keep it like this.